component_sdk/python/kfp_component/_base_op.py

```diff
@@ -0,0 +1,213 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import signal
+import os
+from kubernetes import client, config
+from kubernetes.client.rest import ApiException
+from google.cloud import storage
+import google
+import json
+from datetime import datetime
+import logging
+import sys
+import re
+
+class BaseOp:
+    """Base operation to connect external services.
+    The base class recovers states from persistent storage in the KFP environment.
+    It enables subclasses to resume operation from the recovered state.
+    It handles SIGTERM signal to allow subclasses to cancel operation in case
+    of workflow timeout.
+    Args:
+        name (str):
+            The name of the operation, which will be used as part of the path
+            to store the ```stage_states``` in persistent storage.
+    """
+    def __init__(self, name):
+        self.name = name
+        self.staging_states = {}
+        self._storage = storage.Client()
+        self._staging_bucket = None
+        self._staging_path = None
+        self._argo_workflow_name = None
+        self._argo_node_name = None
+        self._load_k8s_client()
+        self._load_argo_metadata()
+        self._load_staging_location()
+        self._load_staging_states()
+
+    def execute(self):
+        """Executes the operation."""
+        original_sigterm_hanlder = signal.getsignal(signal.SIGTERM)
+        signal.signal(signal.SIGTERM, self._exit_gracefully)
+        try:
+            self.on_executing()
+        except:
+            e = sys.exc_info()[0]
+            logging.error('Failed to execute the op: {}'.format(e))
+            self._stage_states()
+            raise
+        finally:
+            signal.signal(signal.SIGTERM, original_sigterm_hanlder)
+
+    def on_executing(self):
+        """Triggers when execute method is called.
+        Subclass should override this method.
+        """
+        pass
+
+    def on_cancelling(self):
+        """Triggers when the operation should be cancelled.
+        Subclass should override this method.
+        """
+        pass
+
+    def _exit_gracefully(self, signum, frame):
+        logging.info('SIGTERM signal received.')
+        if self._should_cancel():
+            self.on_cancelling()
+        else:
+            self._stage_states()
+
+    def _should_cancel(self):
+        """Checks argo's execution config deadline and decide whether the operation
```

Comment:
Argo doesn't support sending a cancel signal to the container. The only signal it will send is SIGTERM which itself can be triggered from many sources like pod preemption. The signal may be sent in the middle of a retry. The code can still run without argo env, and the cancel feature will be disabled.

Reply:
SG, thanks.